### ISSUE TITLE
A plan for 1.0

### ISSUE THREAD

**Issue by paf31** [purescript/purescript#2805]
I'd like to try to make a plan for 1.0. There are some large changes currently schedules for the 1.0 milestone, so I'll talk about those, and then general goals.

Thoughts on any of this? (please let's discuss specific issues on their respective pages, and keep this discussion general) 

## TL;DR

I would like to cut as many non-essential features as possible in order to get to a stable 1.0 release. What we have now is really quite good, except for a few classes of issues, and I think it is more important and useful to reach stability in terms of the language and compiler features, than it is to add new features or extend existing ones. That said, there are some things we need to do so that the language is stable enough that we don't need to break it for a while after 1.0.

## Planned Changes

In order of decreasing priority (to me):

### Improving error messages

There are a few open issues in 1.0 related to error messages:

- Incorrect error/warning position ranges (#1801)
- Qualify types as necessary in hints in warnings/errors (#1647)
- Add position information to types and data constructors (#2216)
- Show actual code in errors (#1898)
- Type synonyms in error messages (#1821)

I think "Incorrect error/warning position ranges" is probably the most important remaining open non-essential item for 1.0.

### Mark DCE as experimental

`purs bundle` is necessary for development in the browser, as well as the PSCi browser backend and Try PureScript. However, it is slow and involves some heuristics around usages of `require`. A better `purs bundle` would probably operate on one of the two core languages instead of generated JS, but that is out of scope for 1.0. 

For 1.0, I suggest we simply mark dead code elimination as an experimental feature, and continue to use `purs bundle` for bundling in Try PS and PSCi.

### Move to ES6

ES3 vs ES6 is being discussed in #2574 and #2575. I think we should keep the discussion in that ticket, but I'll briefly state my preference here:

> For 1.0, `purs` will continue to use ES3 for both codegen and FFI. Needing to use another transpiler after `purs compile` for any work in the browser just makes ES6 too immature as the default target right now. However, we should aim to get the functional/imperative core into a state where it is possible to develop ES6 support (with its own FFI) as a separate backend without too much effort. Ideally the work on the two core languages will mean that we can share as much of the pipeline as possible between the two backends.

### Imperative core

While not strictly necessary for 1.0, I think it would be good to try to do something to improve the imperative core, and to make it available externally. If this ends up being pushed to post-1.0, I wouldn't be terribly bothered.

### `purs bundle` performance

I would also like to try to improve performance by avoiding parsing if we're not doing dead code elimination. This would probably mean shimming `require` instead of parsing and looking for `require` and then replacing it.

## Other Work

There is some general work which needs to be done before 1.0. This includes things like updating the website and improving documentation in a few places.

**Comment by garyb** [comment]
I guess we should also remove the `class` keyword from imports again, so that we can add constraint kinds as a non-breaking change I think. I can work up a PR that makes it optional again for the remainder of 0.11, and then break it for 1.0?

---

Regarding `purs bundle`, this horse is beaten into microscopic dust, but a few points:

- What we have works - there is no viable alternative yet
- I think referencing its speed is not a great argument for getting rid of it, as I can't possibly imagine a JS replacement will be faster
- The guessword you mention could be called heuristics if you weren't trying to paint it in a bad light :smile: - any other 3rd party tool will have to do the same thing, and re-figure out the rules
- I'm not sure what the mentioned `require` difficulties are, I've written quite a lot of FFI code and never had it behave unexpectedly

So basically we will almost certainly have to copy it as a separate tool, as suggested, unless the rollup plugin starts working or something similar come along before 1.0.

---

All of the error messages stuff depends on the annotations stuff I've been working on, so I'll get that back into shape as soon as the 0.11 chaos is out of the way. :)

**Comment by rvion** [comment]
@garyb if we drop DCE, webpack is a viable and way faster current alternative (incremental bundling, caching, in memory bundle serving, vendors/app bundle splitting , hot module replacement).

I don't know how you set up your live reload/hot reload dev environment on slamdata, but I personally never use purs bundle as it is too slow

**Comment by garyb** [comment]
@rvion I was talking about a viable alternative to using `bundle` to perform DCE. We already use webpack, as we have other JS dependencies that need including via `require`. We just skip running the `bundle` part first for development builds. No hot reloading as it wouldn't play well with the app anyway!

If we don't run DCE the bundle is 10.8mb, with DCE it's 8.4mb. After minification that translates to 4.3mb / 3.4mb, so a 21% reduction in size of what is a pretty chunky file.

As for speed again... it's certainly not the slowest part of the production build either, it runs in ~15s - uglify takes ~50s. That's why you don't use the same process for both dev and production. 😜

**Comment by rvion** [comment]
> I was talking about a viable alternative to using bundle to perform DCE.

Oh, ok, I didn't understand :)

Thanks for all the feedback, though !

> That's why you don't use the same process for both dev and production. 😜

Of course, was only speaking about the bundling phase. Quite a short sighted intervention, though x)

**Comment by anilanar** [comment]
Why don't you use a monorepo approach to better orchestrate release of core libraries?

**Comment by paf31** [comment]
I agree that removing DCE completely is a big change, and you're right that it tends to work as expected. So how about, instead of removing it, we simply mark the DCE feature as experimental for 1.0? Then we could simply try to find a better alternative to the find-and-replace `require` strategy that we have now (perhaps shimming `require`). That's even less work to do 😄 

I'm going to rearrange what I said above to take that into account, and put things in priority order as I see them.

**Comment by paf31** [comment]
@anilanar I've wondered whether a monorepo would work well for the core libraries. We could potentially use `psc-package` to let us develop all packages at once, without the need for `bower link`, so it could be very good for efficiency actually. We should discuss it, but on another issue please.

**Comment by paf31** [comment]
> I guess we should also remove the class keyword from imports again

Will that not cause any problems? If not, and it's not a lot of work, then I think that sounds good.

**Comment by garyb** [comment]
I don't think so, the namespaces already mostly act like they're merged (although @kRITZCREEK mentioned that synonyms might get around it somehow currently).

We did have a period where either way worked before, when we transitioned to not requiring `T()` for type-only imports. It just means the parser can't determine what the import is for, and we need to update it during name desugaring instead.

**Comment by garyb** [comment]
New post looks good to me by the way. :+1:

Also, I think maybe my earlier reply didn't come across in quite the right way - if you think it will be better to remove DCE, then as long as we have extracted a separate DCE-and-bundler and released that when we do 1.0, I think that's an acceptable option. I was mainly saying - what you did with the current `bundle` is really pretty neat, so it's actually set quite a high bar to replace, even if you don't really like the way you did it. :smile:

**Comment by paf31** [comment]
@garyb Since we've got less planned for 1.0 now, maybe we should just include constraint kinds? It'll be useful, and most of the work is done anyway, right?

**Comment by paf31** [comment]
Moving this into the milestone description.
